overlord/overlord.go

```diff
@@ -209,7 +210,30 @@ func (o *Overlord) addManager(mgr StateManager) {
 	o.stateEng.AddManager(mgr)
 }
 
-func loadState(backend state.Backend, restartHandler restart.Handler) (*state.State, error) {
+func initStateFileLock() (*osutil.FileLock, error) {
+	lockFilePath := dirs.SnapStateLockFile
+	if err := os.MkdirAll(filepath.Dir(lockFilePath), 0755); err != nil {
+		return nil, err
+	}
+
+	return osutil.NewFileLockWithMode(lockFilePath, 0644)
+}
+
+func (o *Overlord) loadState(backend state.Backend, restartHandler restart.Handler) (*state.State, error) {
+	flock, err := initStateFileLock()
+	if err != nil {
+		return nil, fmt.Errorf("fatal: error opening lock file: %v", err)
+	}
+	o.flock = flock
+
+	logger.Noticef("Acquiring lock file")
+	// This will cause the process to block, if the lock is currently held
+	if err := o.flock.Lock(); err != nil {
```

Comment:
yes, my preference would still be to actually do some polling and have a timeout after which we fail

Reply:
that would also let us use  sd-notify EXTEND_TIMEOUT_USEC if the systemd timeout is too short

Reply:
I took Ian's suggestion; I actually tried to add a `LockWithTimeout()` method to `flock.go`, but I didn't find a way to cancel out lock when the timeout expires. This would leave the background coroutine running, and eventually (maybe) acquiring the lock, after we already reported an error and gave up.

Reply:
thanks for taking my suggestion, though it's unclear to me if @pedronis wanted the inverse of what I proposed, where we actually will fail at some point if we don't get the lock (i.e. implement a proper timeout), as it is my proposal and this implementation will never actually fail, it will just loop indefinitely trying to acquire the lock and complaining about it every minute.

I personally am okay with the current implementation now, just wondering if Samuele is too

Reply:
I think we agreed elsewhere to switch to a simpler approach using TryLock, which doesn't really need a goroutine or a ticker even

Reply:
I changed it to a retry mechanism. I'm not really happy with the unit tests I wrote, I would rather have an interface for the FileLock, but if this is the preferred way here, so be it. :-)